python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -111,21 +114,36 @@ def _standardizeColumnParameter(self, columns: Dict[str, Union[str, List[str]]])
         name : `str`
             Stringified tuple representing a multi-index column name.
         """
-        if not isinstance(columns, collections.abc.Mapping):
-            raise ValueError("columns parameter for multi-index data frame must be a dictionary.")
-        if not set(self.indexLevelNames).issuperset(columns.keys()):
-            raise ValueError(f"Cannot use dict with keys {set(columns.keys())} "
-                             f"to select columns from {self.indexLevelNames}.")
-        factors = [iterable(columns.get(level, self.columns.levels[i]))
-                   for i, level in enumerate(self.indexLevelNames)]
-        for requested in itertools.product(*factors):
-            for i, value in enumerate(requested):
-                if value not in self.columns.levels[i]:
-                    raise ValueError(f"Unrecognized value {value!r} for index {self.indexLevelNames[i]!r}.")
-            yield str(requested)
-
-    def read(self, columns: Union[str, List[str], Dict[str, Union[str, List[str]]]] = None
-             ) -> pd.DataFrame:
+        if isinstance(columns, list):
+            for requested in columns:
+                if not isinstance(requested, tuple):
+                    raise ValueError(
+                        "columns parameter for multi-index data frame must be either a dictionary or list of tuples."
+                    )
+                yield str(requested)
+        else:
+            if not isinstance(columns, collections.abc.Mapping):
+                raise ValueError(
+                    "columns parameter for multi-index data frame must be either a dictionary or list of tuples."
+                )
+            if not set(self.indexLevelNames).issuperset(columns.keys()):
+                raise ValueError(
+                    f"Cannot use dict with keys {set(columns.keys())} "
+                    f"to select columns from {self.indexLevelNames}."
+                )
+            factors = [
+                iterable(columns.get(level, self.columns.levels[i]))
```

Comment:
Is self.columns always a MultiIndex? From earlier in the code it looks like it can sometimes be an Index which seems not to have a .levels option. I'm not really familiar with pandas and could just be wrong.

I also find it quite confusing to have columns and self.columns which are different things.

Reply:
This whole "else" block I didn't actually change anything from what was previously implemented; I just moved it into an "else" block to be the code.  All I did in this function was to add the `if isinstance(columns, list):` block to handle the case where passing a list of tuples.